### PR TITLE
Integrate the new filter format

### PR DESCRIPTION
**Pull request by AdityaHegde**
## Checklist
- [x] Manual verification
- [ ] Unit test coverage
- [ ] E2E test coverage
- [x] Needs manual QA?

#### Issue addressed: 
To standardise where and having filters we introduced a new format for filters similar to ASTs.

#### Details:
1. Updating the UI to integrate this new format.
2. Moving the code to match the selectors/actions format.

## Steps to Verify
Play around with filters. There shouldnt be any regressions from main.

### CODE REVIEW

File: web-common/src/features/dashboards/filters/Filters.svelte
Comment (bcolloran) [inline web-common/src/features/dashboards/filters/Filters.svelte:177]: i think this can just be `on:click={clearAllFilters}` ?

**Review by bcolloran — CHANGES_REQUESTED** [review]
Hey @AdityaHegde , I didn't have time to do a full deep dive today or to do QA, just a quick pass over the code to get a sense of the shape of it. From what I understand of it so far, I think it's looking good. I'll try to go deeper Monday.

File: web-local/test/ui/utils/dashboardHelpers.ts
Comment (bcolloran) [inline web-local/test/ui/utils/dashboardHelpers.ts:130]:
```suggestion
  const includeFilterRequest = new Map<string, string[]>();
```

File: web-local/test/ui/utils/dashboardHelpers.ts
Comment (bcolloran) [inline web-local/test/ui/utils/dashboardHelpers.ts:131]: as above
```suggestion
  const excludeFilterRequest = new Map<string, string[]>();
```

**Review by bcolloran — CHANGES_REQUESTED** [review]
Great PR @AdityaHegde -- there is some complicated stuff in here, but it's handled with nice clarity and nice abstractions, and overall I think there is actually a lot of simplification here even with the added functionality.

I have a bunch of change requests, but it's really nothing major because I think the overall shape is really good. It's mostly just small TS nits to bring the code a little more inline with TS best practices (from e.g. the Google TS style guide and elsewhere https://google.github.io/styleguide/tsguide.htm). There are no large change requests, so hopefully this will not take long, but it's a big PR so there are a lot of requests...

Last thing: this was a big one and took me almost all day, but I still didn't have time to do a QA pass. Is there good e2e coverage for this functionality? If not, we should really add that before merging. If there is, then a QA pass is probably not really needed.

File: web-common/src/features/dashboards/stores/filter-utils.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/stores/filter-utils.ts:171]: Hmmm i recently fixed an issue with non string dimensions. We should update the type everywhere and keep this as any for now.

File: web-common/src/features/dashboards/dimension-table/dimension-table-utils.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/dimension-table/dimension-table-utils.ts:47]: It is not needed. It can be assumed the caller has created a copy for this dimension.

File: web-common/src/features/dashboards/dimension-table/dimension-table-utils.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/dimension-table/dimension-table-utils.ts:58]: Adding `%` is a UX decision right now. It feels cleaner for the caller to make that distinction IMO.

File: web-common/src/features/dashboards/proto-state/toProto.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/proto-state/toProto.ts:251]: Sounds good. Updated to force convert using `JSON.stringify`. We should do a global refactor of dimension value to be anything and not just string. This can hopefully be fixed then.

File: web-common/src/features/dashboards/proto-state/toProto.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/proto-state/toProto.ts:218]: It can actually be anything. Ties into a greater refactor of dimension values.

File: web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:47]: Let this be here for now. We need a greater refactor to handle non string dimension values.

File: web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts
Comment (AdityaHegde) [reply web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:104]: Is this how typescript recommends it? I remember sonarqube giving the way i have written a while back. (It was configured by a java dev where this makes sense so i am not sure)

Comment (AdityaHegde) [reply web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:104]: It also matches `new Set<string>()` and `new Map<string, ...>()` declaration-assignments.

**Comment by AdityaHegde** [comment]
Thanks for the review. There is an issue right now with us assuming dimension values are string. This resulted in a few issues. But the type declarations for dimension value params and other places were not updated. We should go through and get a consistent typing for it. Hence, I have left in the any in the filter utils for now.

Regarding E2E, there is a test that changes the filters. There were some failures around it while developing so it catches issues there I believe.

File: web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts
Comment (bcolloran) [reply web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:47]: Ok sounds good. In that case, can you leave a FIXME with a note about the cast? -- the reason for casting, and the fact that we want to remove it in future work

File: web-common/src/features/dashboards/dimension-table/dimension-table-utils.ts
Comment (bcolloran) [reply web-common/src/features/dashboards/dimension-table/dimension-table-utils.ts:55]: sounds good!

File: web-common/src/features/dashboards/proto-state/fromProto.ts
Comment (bcolloran) [inline web-common/src/features/dashboards/proto-state/fromProto.ts:184]: Here's a little trick for filters like this @AdityaHegde -- did you know that you can use:
`.filter((e): e is V1Expression => e !== undefined)`
to avoid the cast? It basically asserts that if `e is V1Expression` then the predicate must be true, and then TS is smart enough to propagate the type! I only just learned this the other day myself (I have code that I'll need to clean up to use this), so I thought I'd share it :-)

File: web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts
Comment (bcolloran) [reply web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts]: Yep I think this is the recommendation these days -- even Google, who recommends the `Array<T>` notation for complicated types would suggest the shorthand notation here
https://google.github.io/styleguide/tsguide.html#arrayt-type

File: web-common/src/features/dashboards/stores/filter-utils.ts
Comment (bcolloran) [reply web-common/src/features/dashboards/stores/filter-utils.ts]: I really don't think this should be any. If it's not `string`, then I think it must be something like `string | number`. @begelundmuller -- can you weigh in here? If we've recently encountered error with non-string dimensions, can you help us out by telling us the complete universe of possible dimensions we might encounter?

FWIW, part of the reason I'm pushing on this is because the work on the downstream component code, for example all the stuff I did in the leaderboards last month, is made way more complicated if I can't have guarantees about the shape of the data used by the UI. It's a nightmare if we have to guard against everything that close to the display layer, and we should narrow these types closer to the data layer.

**Comment by begelundmuller** [comment]
@bcolloran The protobuf type is `google.protobuf.Value` which can be any value that serializes to JSON. The value is propagated more or less directly to a DuckDB query arg (so e.g. a nested object might be valid if the underlying DuckDB type is a struct).

**Comment by bcolloran** [comment]
@AdityaHegde please go ahead and merge this when you are ready. I'm sorry, I should have made more clear: I think this PR is ready to go and it's great! I wanted to dig more into the typing, but I didn't mean to hold you up

> @bcolloran The protobuf type is `google.protobuf.Value` which can be any value that can serialize to JSON. The value is propagated more or less directly to a DuckDB query arg (so e.g. a nested object might be valid if the underlying DuckDB type is a struct).

@begelundmuller ok awesome, thanks that's super helpful! Would it be possible to add a comment that explains why this needs be `unknown` (i.e. that it could represent any supported DuckDB expression value, including possibly things like nested structs etc)? I think you said it's easy to add comments in the API code that are passed through the codegen to the TS types, right?

**Pull request merged**
